apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/SpannerException.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Linq;
+using Google.Cloud.Spanner.V1.Logging;
+using Grpc.Core;
+
+namespace Google.Cloud.Spanner
+{
+    /// <summary>
+    /// Represents an error communicating with the Spanner database.
+    /// </summary>
+    public sealed class SpannerException : Exception
+    {
+        /// <summary>
+        /// This class is a thin conversion around a grpc exception, with the additional
+        /// information of whether the operation is retryable based on the resulting error.
+        /// </summary>
+        /// <param name="code"></param>
+        /// <param name="innerException"></param>
+        internal SpannerException(ErrorCode code, RpcException innerException)
+            : base(GetMessageFromErrorCode(code), innerException)
+        {
+            Logger.LogPerformanceCounterFn("SpannerException.Count", x => x + 1);
+            ErrorCode = code;
+        }
+
+        internal SpannerException(ErrorCode code, string message)
+            : base(message)
+        {
+            Logger.LogPerformanceCounterFn("SpannerException.Count", x => x + 1);
+            ErrorCode = code;
+        }
+
+        internal SpannerException(RpcException innerException)
+            : this(ConvertFromStatusCode(innerException.Status.StatusCode), innerException)
+        {
+        }
+
+        /// <summary>
+        /// </summary>
+        public ErrorCode ErrorCode { get; }
+
+        /// <summary>
+        /// </summary>
+        public bool IsRetryable
+        {
+            get
+            {
+                switch (ErrorCode)
+                {
+                    case ErrorCode.DeadlineExceeded:
+                    case ErrorCode.Aborted:
+                    case ErrorCode.Unavailable:
+                        return true;
+                    default:
+                        return false;
+                }
+            }
+        }
+
+        internal static bool TryTranslateRpcException(Exception possibleRpcException,
```

Comment:
Pattern-based switch will make this much nicer :)
You might want to consider a "null means no translation" instead of using the out parameter - see whether it makes things simpler or not. (You can do so because you'd never set spannerException to null *and* return true.)